quint/src/effects/inferrer.ts

```diff
@@ -192,14 +193,17 @@ export class EffectInferrer implements IRVisitor {
         const substitution = arrowEffect.chain(effect => unify(signature, effect))
 
         const resultEffectWithSubs = substitution
-          .chain(s => compose(s, this.substitutions))
+          .chain(s => compose(this.substitutions, s))
           .chain(s => {
             this.substitutions = s
 
-            this.effects.forEach((effect, id) => {
-              const r = applySubstitutionToScheme(s, effect)
+            paramsResult.map(effects => zip(effects, expr.args.map(a => a.id)).forEach(([effect, id]) => {
+              if (!effect || !id) {
+                throw new Error('Invalid arrays')
```

Comment:
This error seems a bit cryptic. Is it meant to be an impossible case? Is the actual problem here that we don't have proper paring between effects and ids? Maybe the message could indicate what is wrong with the arrays.

Reply:
It's an impossible scenario: it would happen if `effects` and `expr.args` had different sizes, but they can't. I'll add a comment and improve the message :smile: